### ISSUE TITLE
Segfault when running nn.StochasticGradient training

### ISSUE THREAD

**Issue by frkl** [torch/torch7#261]
Ran into a segfault when trying to train a Lenet on mnist. Training crashes randomly after 0-2 epoches in different runs.

=============code=============
net = nn.Sequential()
net:add(nn.SpatialConvolution(1, 6, 5, 5))
net:add(nn.SpatialMaxPooling(2,2,2,2)) 
net:add(nn.SpatialConvolution(6, 16, 5, 5))
net:add(nn.SpatialMaxPooling(2,2,2,2))
net:add(nn.View(16*5*5))
net:add(nn.Linear(16*5*5, 120))
net:add(nn.Linear(120, 84))
net:add(nn.Linear(84, 10))   
net:add(nn.LogSoftMax())

criterion = nn.CrossEntropyCriterion()
trainer = nn.StochasticGradient(net, criterion)
trainer.learningRate = 0.001
trainer.maxIteration = 5

trainer:train(mnist_train_dataset)

=============output=============
 StochasticGradient: training	
 current error = 0.6249592341091	
 current error = 0.54681050086858	
 *** Error in `/home/****/torch/install/bin/luajit': free(): invalid next size (fast): 0x00000000017b4ed0 ***
 ...
I'm running with Ubuntu 14.04 on a VM with 2GB memory.

**Comment by ajtulloch** [comment]
Did this get resolved? It sounds like a possible OOM, especially with a 2GB VM? I'd try on a VM with more memory (8GB+) .

**Comment by frkl** [comment]
I reduced the learning rate from 0.001 to 0.0001 and it worked.

It seemed like nan weights during optimization caused the problem. There seem to be 3 crashing spots in the nn (not torch7 it seems) package when the net is producing nan output during training: 

nn/StochasticGradient.lua line 37:
       module:updateGradInput(input, criterion:updateGradInput(module.output, target))

nn/Sequential.lua line 93:
      currentModule:accUpdateGradParameters(previousModule.output, currentGradOutput, lr)

nn/Sequential.lua line 98:
   currentModule:accUpdateGradParameters(input, currentGradOutput, lr)
